### ISSUE TITLE
Why Request.HttpMethod is enum ? Is it possible to make it as a String ?

### ISSUE THREAD

**Issue by glani** [OpenFeign/feign#994]
The reasonable question why it has been implemented as enum ? Feign is a good tool to use for communication with webdav. However webdav uses another official standard(https://tools.ietf.org/html/rfc4918) and requires to pass instead of pre-defined GET, POST etc their owns  method names.  Is it possible to make it as a String ?

I implemented some hacks to extend the enum via reflection. It works fine with okhttp and does not work with default client (there is the same restriction for method name value). apache http client should work fine as well because org.apache.jackrabbit:jackrabbit-webdav uses it as a basis. Anyway Feign allows us to change http client and use our own however there have been this restriction.

Thank you very much in advance.

**Comment by velo** [comment]
wow, serious?!

I had no idea about that. If I'm not mistaken, we introduced this on feign 10.

Well, I would say that would mean reverting that change.
Or making something similar to that JDK does for `CopyOption`.

Could you share your hack in case anyone else also subs onto this?!

**Comment by kdavisk6** [comment]
We could also update the enum to include those options too.

**Comment by velo** [comment]
good point! How many verbs are we talking about for webdav?

**Comment by glani** [comment]
> Could you share your hack in case anyone else also subs onto this?!

Easy:

// Shared part
{
public enum DavMethodsValues {

    PROPFIND,

    PROPPATCH,

    MKCOL,

    COPY,

    MOVE,

    LOCK,

    UNLOCK,

    ORDERPATCH,

    SUBSCRIBE,

    UNSUBSCRIBE,

    POLL,

    SEARCH,

    REPORT,

    VERSION_CONTROL("VERSION-CONTROL"),

    CHECKIN,

    CHECKOUT,

    UNCHECKOUT,

    LABEL,

    MERGE,

    UPDATE,

    MKWORKSPACE,

    BASELINE_CONTROL("BASELINE-CONTROL"),

    MKACTIVITY,

    ACL,

    REBIND,

    UNBIND,

    BIND;

    private final String realValue;

    DavMethodsValues() {
        this(null);
    }

    DavMethodsValues(final String realValue) {

        this.realValue = realValue;
    }

    public String getRealValue() {
        return realValue != null ? realValue : name();
    }
}


// java 11

    // original answer
    // https://stackoverflow.com/questions/9614282/how-to-create-an-instance-of-enum-using-reflection-in-java/51244909#51244909
    public static void setupEnums() throws Throwable {
        final Constructor<?> constructor = Unsafe.class.getDeclaredConstructors()[0];
        constructor.setAccessible(true);

        final Field ordinalField = Enum.class.getDeclaredField("ordinal");
        makeAccessible(ordinalField);

        final Field nameField = Enum.class.getDeclaredField("name");
        makeAccessible(nameField);

        final Field valuesField = Request.HttpMethod.class.getDeclaredField("$VALUES");
        makeAccessible(valuesField);

        final Request.HttpMethod[] oldValues = (Request.HttpMethod[]) valuesField.get(null);
        final Request.HttpMethod[] newValues = new Request.HttpMethod[oldValues.length + DavMethodsValues.values().length];
        System.arraycopy(oldValues, 0, newValues, 0, oldValues.length);


        int ordinal = Request.HttpMethod.values().length;
        final Unsafe unsafe = (Unsafe) constructor.newInstance();
        for (final DavMethodsValues value : DavMethodsValues.values()) {
            final Request.HttpMethod enumValue = (Request.HttpMethod) unsafe.allocateInstance(Request.HttpMethod.class);

            ordinalField.setInt(enumValue, ordinal);
            nameField.set(enumValue, value.getRealValue());
            newValues[ordinal] = enumValue;
            ordinal++;
        }
        valuesField.set(null, newValues);
    }

    private static void makeAccessible(final Field field) throws Exception {
        field.setAccessible(true);
        final Field modifiersField = Field.class.getDeclaredField("modifiers");
        modifiersField.setAccessible(true);
        modifiersField.setInt(field, field.getModifiers() & ~ Modifier.FINAL);
    }


// Java 8 (EnumBuster from https://www.javaspecialists.eu/archive/Issue161.html):

    public static void setupEnums() throws Throwable {
        final EnumBuster<Request.HttpMethod> buster = new EnumBuster<>(Request.HttpMethod.class);
        for (final DavMethodsValues value : DavMethodsValues.values()) {
            final Request.HttpMethod v = buster.make(value.getRealValue());
            buster.addByValue(v);
        }
    }

Unfortunately there has been discussion in the team to get rid of using feign for both webdav and eureka . We found one critical issue: the feign is not a stream oriented .Request.Body is final and always requires an array of bytes. If api requires to send 1 gb file to webdav or between services we do not have to use feign and write our own wrapper upon okhttp. I found a discussion about whether feign should support IO Stream (~2015) as I understand there have been no changes and no final decision. Despite the fact that the position of the contributors that is light-weight http simple wrapper , unfortunately the world is not only to create, check and retrieve token apis but also to transfer data and sometimes a huge amount.

**Comment by glani** [comment]
> We could also update the enum to include those options too.

I checked how it has been done in okhttp. They use method check for body as a string with description how it has to be done in webdav. From my point of view having such things hardcoded and fixed is not good. It would be better if you did as a string.

**Comment by kdavisk6** [comment]
@glani 

There are some drawbacks to using Strings.  One major one is that it complicates our ability to be proactive when validating an interface against our contract, as our list of **known good** values is essentially everything.  This type of flexibility is detrimental to maintenance as it requires the use of **magic strings** to be used during validation, which is always flaky and bad practice, hence the enum.

We feel it is better to catch these errors before the Feign Target is created rather than allowing anything to be sent and having the requests fail or the Error Decoder be called because of an invalid method definition.

There isn't much risk in *hard-coding* these values using an enum, as the specifications we are conforming to [RFC 7231](https://tools.ietf.org/html/rfc7231#section-4.3) and [RFC 4918](https://tools.ietf.org/html/rfc4918) are well established and not likely to change without notice.

While there is some impact to those extending Feign and writing their own contracts, that impact would occur anyway, as the contract interface requires implementations to manage their own annotations.

Simply put, we see the specificity as a benefit.

**Comment by glani** [comment]
@kdavisk6 fine. How do you want to act with methods that contain "-" in names inside? I mean that in the contract enumvalue.name() is used. Reflection works fine with it, drawbacks are that it is necessary to support different jdks and warning about illegal reflection.
